SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,471 @@
+	FSpatialGDKSpanId NewSpanId = EventTracer->TraceEvent(FSpatialTraceEventBuilder::CreateSendRPC(LinearTraceId),
+														  /* Causes */ SpanId.GetConstId(), /* NumCauses */ 1);
+
+	if (OutUpdates.Num() == 0 || OutUpdates.Last().EntityId != EntityId || OutUpdates.Last().Update.component_id != ComponentId)
+	{
+		OutUpdates.AddDefaulted();
+		OutUpdates.Last().EntityId = EntityId;
+		OutUpdates.Last().Update.component_id = ComponentId;
+	}
+	OutUpdates.Last().Spans.Add(NewSpanId);
+}
+
+void USpatialNetDriverRPC::OnDataWritten(TArray<FWorkerComponentData>& OutArray, Worker_EntityId EntityId, Worker_ComponentId ComponentId,
+										 Schema_ComponentData* InData)
+{
+	if (ensure(InData != nullptr))
+	{
+		FWorkerComponentData Data;
+		Data.component_id = ComponentId;
+		Data.schema_type = InData;
+		OutArray.Add(Data);
+	}
+}
+
+void USpatialNetDriverRPC::OnUpdateWritten(TArray<UpdateToSend>& OutUpdates, Worker_EntityId EntityId, Worker_ComponentId ComponentId,
+										   Schema_ComponentUpdate* InUpdate)
+{
+	if (ensure(InUpdate != nullptr))
+	{
+		if (OutUpdates.Num() == 0 || OutUpdates.Last().EntityId != EntityId || OutUpdates.Last().Update.component_id != ComponentId)
+		{
+			OutUpdates.AddDefaulted();
+			OutUpdates.Last().EntityId = EntityId;
+			OutUpdates.Last().Update.component_id = ComponentId;
+		}
+		OutUpdates.Last().Update.schema_type = InUpdate;
+	}
+}
+
+USpatialNetDriverRPC::StandardQueue::SentRPCCallback USpatialNetDriverRPC::MakeRPCSentCallback(TArray<UpdateToSend>& OutUpdates)
+{
+	if (EventTracer != nullptr)
+	{
+		return [&OutUpdates, this](FName Name, Worker_EntityId EntityId, Worker_ComponentId ComponentId, uint64 RPCId,
+								   const FSpatialGDKSpanId& SpanId) {
+			return OnRPCSent(EventTracer, OutUpdates, Name, EntityId, ComponentId, RPCId, SpanId);
+		};
+	}
+	return StandardQueue::SentRPCCallback();
+}
+
+RPCCallbacks::DataWritten USpatialNetDriverRPC::MakeDataWriteCallback(TArray<FWorkerComponentData>& OutArray) const
+{
+	return [&OutArray](Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentData* InData) {
+		return OnDataWritten(OutArray, EntityId, ComponentId, InData);
+	};
+}
+
+RPCCallbacks::UpdateWritten USpatialNetDriverRPC::MakeUpdateWriteCallback(TArray<UpdateToSend>& OutUpdates) const
+{
+	return [&OutUpdates](Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentUpdate* InUpdate) {
+		return OnUpdateWritten(OutUpdates, EntityId, ComponentId, InUpdate);
+	};
+}
+
+void USpatialNetDriverRPC::Init(USpatialNetDriver* InNetDriver, const SpatialGDK::FSubView& InActorAuthSubView,
+								const SpatialGDK::FSubView& InActorNonAuthSubView)
+{
+	NetDriver = InNetDriver;
+	LatencyTracer = USpatialLatencyTracer::GetTracer(InNetDriver->GetWorld());
+	EventTracer = InNetDriver->Connection->GetEventTracer();
+	RPCService = MakeUnique<SpatialGDK::RPCService>(InActorNonAuthSubView, InActorAuthSubView);
+
+	{
+		// TODO
```

Comment:
remove?

Reply:
This class is a skeleton to be fleshed out later. I would keep that. I'll add the UNR ticket.

Reply:
Personally I would not submit this, you can keep it locally however.